[MU-2] Add Parameter To init To Allow Selection of File Types

knittingarch opened the pull request:
This is a WIP Media Uploader widget.  It uses JS, JQuery, FileReader, and Underscore only.

This version replicates (with one deviation) the existing format of this widget.  There is a preview of the uploaded file as well as data related to that file at the same time.  There are also a few stylistic changes.

_N.B. Currently you need to manually input the file type in the init, `mediaUploader.init()`, to populate the templates correctly. This will driven off of the media menu-bar in CommCare in the future._

**Incomplete Tasks**
- The `View Full Size` button does not currently work.  There is an issue with the URL data exceeding the allowed limit.  This will require more research.
- The upload button has several states that have not yet been implemented.
- There are several alerts that appear and divs that disappear based on the upload status.  That functionality still needs to be created.
- All of this will need to be integrated into CommCare (which may solve the button issue referenced above.)

knittingarch committed:
Add Click Event For Full Size Image Preview

This commit creates the mechanism to show full size image previews.  The functionality is not quite there yet, but it should be simple to swap out the code that is there for code that does what is needed.

knittingarch committed:
Clean Up Code And Add Another Temp Image

knittingarch committed:
Add Parameter to Init

This commit adds a parameter to the init as the preliminary step towards dynamically selecting a media type.  The next step will be to add a dictionary of items that can be indexed manually.  After that a dynamic implementation will need to be created.

knittingarch committed:
Add Dictionary To Class And Index It For Function Params

This commit adds the dictionary with text and glyphicon for each file type.  It also indexes this dictionary in order to populate the required params in the subsequent functions.  This is still not dynamic, but a good step towards that implementation.

knittingarch committed:
Clean Up Code And Add Missing Styling

knittingarch committed:
Clean Up CSS and Restructure Code

This commit cleaned up most of the remaining unused CSS as well as reordered some of the divs to make the code flow in a more logical way.

orangejenny commented:
@millerdev @emord @biyeun

knittingarch committed:
Add Full Size Image Event And Conditional

This commit sets up the nfrastructure for a future function that will show the full sized image in a new window.  Currently it is not possible to use the selected image as it has to be saaved to the server first.  Hopefully once this code is integrated, it will be possible to use this code effectively once it is integrated into CommCare.

biyeun commented on uploader.html:
```diff
@@ -178,96 +175,88 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
-      // private IDs and Classes 
+      self.init = function (filetype) {
+        var mediaTypes = [
+          {'image': {'text': 'image', 'glyphicon': 'glyphicon glyphicon-picture'}, 'audio': {'text': 'audio', 'glyphicon': 'glyphicon glyphicon-volume-up'}, 'video': {'text': 'video', 'glyphicon': 'glyphicon glyphicon-facetime-video'}, 'video-inline': {'text': 'video-inline', 'glyphicon': 'glyphicon glyphicon-play'}, 'expanded-audio': {'text': 'expanded-audio', 'glyphicon': 'glyphicon glyphicon-v ...
```
Is there a reason why this is a single dict inside of a list? Also, is there a reason we're using glyphicons here when the rest of the site is using the fontawesome library? @orangejenny was this part of the requirements?
 Also while it will take up more vertical space, it makes it much easier to follow for another developer and usually better practice to format dicts as:
```
var mediaTypes = {
    image : {
        text: 'Image',
        icon: '....',
    },
    audio: {
        text: 'Audio',
        icon: '...',
    },
    ...
};

orangejenny replied:
I just forgot. Sarah, check out [fontawesome's icon library](http://fontawesome.io/cheatsheet/)

knittingarch replied:
Changed!

biyeun commented on uploader.html:
```diff
@@ -178,96 +175,88 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
           {fileName: fileName,
-           fileSize: fileSize,
-           filetype: "image"}
+           fileSize: fileSize}
```
I know the original code wasn't following our current best practices, but it should be...
```
$(mediaData).html(dataTemplate({
    fileName: fileName,
    fileSize: fileSize,   
}));
```
Yes, the last trailing comma is intentional. ;-)

knittingarch replied:
Done!

biyeun commented on uploader.html:
```diff
@@ -178,96 +175,88 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
       var self = {},
```
before this line:
```
'use strict';
```

knittingarch replied:
Done!

biyeun commented on uploader.html:
```diff
@@ -147,29 +144,29 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
+          </div><!-- close modal-body -->
 
           <div class="modal-footer">
+            <input type="button" value="Close" class="btn btn-sm btn-default" data-dismiss="modal">
```
I know that the old code was doing this, but now we're using `<button>` tags instead of `<input type="button" />` because it makes adding translations and icons easier.
```
<button type="button" class="btn btn-sm btn-default data-dismiss="modal">Close</button>
```

knittingarch replied:
I wondered about this. Does this apply to all buttons (ie. submit)?

knittingarch replied:
@biyeun Can you take a look at this?  I'd like to close out this PR this week, if possible.

biyeun commented on uploader.html:
```diff
@@ -147,29 +144,29 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
             <!-- selection section container -->
             <div id="js-selected-media-section">
               <script type="text/html">
```
it's better practice to apply the template `id` to the `<script>` tag itself. And even better to put all templates together in their own section that's removed from the rest of the view (not nested in the `modal`). I know the original code wasn't following the best practices

knittingarch replied:
I like the idea of separating the functional from the stylistic, but I worry about having extra classes lying around.  

I also am not entirely sure I understand the "put all the templates in their own section" suggestion.  Would this entail creating another div id to replace the one that is now being assigned to the style tag?

knittingarch replied:
I was able to figure this out without adding additional classes.  Those changes can be found on the [media-uploader-sarah](https://github.com/dimagi/MediaUploader/tree/media-uploader-sarah) in the uploader.html file.

biyeun commented on uploader.html:
```diff
@@ -18,73 +18,70 @@
       padding-bottom: 20px;
-      /*background-color: lightsalmon;*/
     }
     #js-output-file {
```
it's generally better practice to give a separate styling class than the id referenced in javascript

knittingarch replied:
Done!

biyeun commented:
yoooo but that unicorn...

millerdev commented on uploader.html:
```diff
@@ -178,96 +175,88 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
 
-      // init
-      self.init = function () {
         $(document).ready(function() {
```
Should be `$(function () {`

millerdev commented on uploader.html:
```diff
@@ -178,96 +175,88 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
-          createModalHeader();
+          createModalHeader(mediaType['text']);
+        });
+        $(document).on('click', opts.selectFileSelector, function() {
```
Should be `$(opts.selectFileSelector).click(function() {`

millerdev commented on uploader.html:
```diff
@@ -178,96 +175,88 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
+          _private.createSelectFileSection(mediaType['text']);
         });
-        $(document).on('click', opts.selectFileSelector, _private.createSelectFileSection);
         $(document).on('change', opts.fileFieldSelector, function(event) {
```
Should be `$(opts.fileFieldSelector).change(function(event) {`

orangejenny replied:
Why is that? I believe Sarah's using `on` here because `$(opts.fileFieldSelector)` may not exist at the time this code runs.

millerdev replied:
oh, that makes sense. I hadn't thought about it not existing.

millerdev commented on uploader.html:
```diff
@@ -178,96 +175,88 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
 
       // get data from file selector
       _private.collectFileData = function(event) {
+        var image = {};
```
curious about why this was moved out here. The previous way seems better since (as far as I can see) it's only referenced inside the `onload` handler.

knittingarch replied:
I was hoping to use it elsewhere, but removed that code because it didn't work and muddied the testing.

knittingarch replied:
@millerdev Can you take a look at this?  Hoping to get this PR closed this week :)

millerdev commented on uploader.html:
```diff
@@ -283,7 +272,7 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
         readyTextSelector: "#js-file-ready-text",
       });
-      mediaUploader.init();
+      mediaUploader.init('image');
```
Curious why this `init(...)` call is separate from the constructor? Is it intended to be called multiple times per `MultimediaUploader` instance? If so, I'd suggest renaming it to something like `addFileType(type)`. If not, I'd suggest putting the filetype(s) in a property of the options passed into the constructor and make the `init` code private/called by the constructor.

knittingarch replied:
Not entirely sure I understand this comment.  Can you provide a little more description of the desired change?

millerdev replied:
https://github.com/knittingarch/MultimediaUploader/pull/1/files/bb8bf535f1146b52a7b1ebdb1552ec649e140ac4#r140398360

millerdev reviewed (commented):
That unicorn made my day!

orangejenny commented on uploader.html:
```diff
@@ -18,73 +18,70 @@
     }
     .file-data,
-    .file-progress,
+    .progress-bar,
```
Is this using bootstrap's progress bar? If so, I wouldn't expect you'd need to override their `.progress-bar` styles.

knittingarch replied:
I've removed this, but the progress bar now has different styling than the surrounding divs.

orangejenny commented on uploader.html:
```diff
@@ -18,73 +18,70 @@
-    progress[value] {
-      height: 30px;
+    .progress-bar {
+      width: 359px;
```
Is there a reason this is a static width? Those generally don't scale well to larger/smaller screens.

knittingarch replied:
If I don't specify a width, the progress bar doesn't show up at all.  I tried various percentages, but they didn't quite match the div above.

<img width="672" alt="screenshot 2017-09-21 12 16 06" src="https://user-images.githubusercontent.com/7423975/30706579-c1d7db78-9ec6-11e7-9c10-744efdfac847.png">

orangejenny commented on uploader.html:
```diff
@@ -18,73 +18,70 @@
+      max-width: 150px;
+      max-height: 150px;
+    }
+    .btn-default {
```
Why override the default bootstrap styles?

knittingarch replied:
Agreed.  There is some weirdness that is making these overrides necessary.  Love to get to the bottom of this one and the progress bar issue.

orangejenny commented on uploader.html:
```diff
@@ -18,73 +18,70 @@
+      line-height: 7;
+    }
+    .modal-footer {
+      padding-right: 31px;
```
This looks oddly specific, what's it for?

knittingarch replied:
I was using this to align the right edge of the Begin Upload button and the data divs after image selection.  I've removed it.

orangejenny commented on uploader.html:
```diff
@@ -111,8 +107,8 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
             <div id="select-file-section" class="row"></div>
-            <div id="media-data-section" class="row col-sm-8 pull-right"></div>
             <div id="selected-media-section" class="row col-sm-12"></div>
+            <div id="media-data-section" class="row col-sm-8 pull-right"></div>
```
Is this meant to be 8 columns wide, but right-aligned instead of left-aligned? If so, should use `col-sm-offset-4` instead of `pull-right`.

knittingarch replied:
Done!

orangejenny reviewed (commented):
A couple of overall thoughts
- Can you add a screenshot (or a few)?
- Split the JavaScript and styling into separate files
- My fault for not asking for this in the first place, but it'd be good to move this into dimagi's mediauploader repository, both because the notifications will work better (most people have dimagi repos only set up to notify their work email addresses) plus then our JavaScript linter will run. The most straightforward way to do this is probably to commit uploader.html in its current state to the dimagi repo in one commit, and then handle PR feedback in subsequent commits, so reviewers can easily see what's changed since they reviewed. Would also be good to make it a PR against a non-master branch for now, so that you can create and merge this and future PRs but your code won't be pulled down by HQ (which depends on `master`) until it's ready.

knittingarch committed:
Address Some Of The PR Concerns

This commit covers the following issues about 60% of the issues identified by reviewers.  Specifically, code standards were applied and glyphicons were switched out for font awesome icons.  There are a few requested changes that are unclear and I will follow up with the reviewers individually.

knittingarch commented:
Initial View Before File is Selected

https://photos-4.dropbox.com/t/2/AADJV7qra1Nl5JlqVydZ8Y1WhJAq9z6qCzbLa1vt4xFNmQ/12/18996597/png/32x32/1/_/1/2/Screenshot%202017-09-21%2011.52.47.png/EKSjmQ4YoMQ5IAcoBw/WdO2ywT1j6rlpN0GEK7lrJnXvAmGA9jEbHNgbTG67ak?size=1280x960&size_mode=3

knittingarch commented:
View Once File is Selected
https://photos-3.dropbox.com/t/2/AAAqxGK_PaYgXkcfCYu-nlGu7benbV33ump9k45cwupjgw/12/18996597/png/32x32/1/_/1/2/Screenshot%202017-09-21%2011.53.06.png/EKSjmQ4YoMQ5IAcoBw/sRNjTF0yryrJh3at_inF8JGxLIrSe92hYhhzJcE47mY?size=1280x960&size_mode=3

knittingarch committed:
Split Out JS Code Into Separate File

knittingarch commented:
@biyeun @millerdev @orangejenny  This code has moved to https://github.com/dimagi/MediaUploader/tree/media-uploader-sarah.  I will continue to check here for feedback, but code changes will be made on the new branch.

millerdev commented on uploader.js:
```diff
@@ -0,0 +1,135 @@
+        progressBarSelector: "#js-file-progress",
+        readyTextSelector: "#js-file-ready-text",
+      });
+      mediaUploader.init('image');
```
What I had in mind was removing this line and adding an extra option to be passed into the contructor:
```js
var mediaUploader = new MultimediaUploader({
  filetype: 'image',
  selectHeaderSelector: "#js-create-header",
  ...
});
```
The `init` function should be made private and called by the constructor:
```js
  ... // inside MultimediaUploader

  _private.init = function (filetype) {
    ... // init code
  };
  _private.init(ops.filetype);

  ... // other private methods

  return self;
}
```
Technically the `init` code could be inlined directly inside `MultimediaUploader` if you want, but it might be cleaner to keep it in a private function—your choice.

knittingarch replied:
Makes sense. Thanks for the example code!

knittingarch committed:
Add New Repo Code Changes Back To Personal Repo

This commit moves the changes that currently reside on dimagi/MediaUploader back to this repo in order to facilitate a smoother conclusion to the PR.

knittingarch commented:
False alarm!  I've moved the changes back to this repo since it was confusing having code that was being updated in two places.  I'll make all remaining changes here and then copy these files back over to the dimagi/MediaUploader repo.

orangejenny commented on uploader.html:
```diff
@@ -110,183 +89,76 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
           <div class="modal-footer">
-            <!-- <span class="glyphicon glyphicon-cloud-upload"></span> -->
+            <button type="button" class="btn btn-sm btn-default" data-dismiss="modal">Close</button>
+            <!-- <i class="fa fa-cloud-upload"></i> -->
```
dead code, should either uncomment or remove

knittingarch replied:
Actually, I meant to move this into the `button` tag so that the icon shows on the button.  Have done that now.

orangejenny commented on uploader.html:
```diff
@@ -110,183 +89,76 @@ <h4 class="modal-title">Upload <%= filetype %></h4>
-  <script>
-    var MultimediaUploader = function(opts) {
-      'use strict';
+        </div><!-- close modal-content -->
```
thank you, these comments are a handy convention

knittingarch replied:
👍

orangejenny reviewed (approved):
Looks good to me, ready to move into a branch in dimagi/MediaUploader once Daniel and Biyeun's comments are addressed.

millerdev reviewed (commented):
really sorry I didn't post this earlier. I wrote up the answer to your question a few days ago, but forgot that it was part of an unsubmitted review. :(

millerdev commented on uploader.js:
```diff
@@ -0,0 +1,135 @@
+          console.log(image);
+          window.open('http://www.dogster.com/wp-content/uploads/2016/07/shutterstock_90574015-600x368.jpg');
+        } else {
+          alert('no image');
```
`console.log` and `alert` are typically used for debugging only. I'm assuming you'll remove these when you move the code into dimagi/MediaUploader, but just wanted to flag.

knittingarch replied:
Ah, good catch.  I'll remove those.

knittingarch replied:
Ah, so it looks like I didn't make the changes, but I have just made them on the new branch.  I forgot that there was weirdness because I moved things over to MediaUploader after the initial PR reviews.  So, I'm not sure what Jenny will want to do with this PR, but the new and improved code can be found here: https://github.com/dimagi/MediaUploader/compare/media-uploader-sarah

knittingarch commented:
@millerdev Does this have your approval?

millerdev reviewed (commented):
are you still planning to address the comment I posted last time? I didn't see an update related to that. Did I miss it? Also one other small thing noted below.

millerdev reviewed (approved):
cool. that other branch looks fine to me
